### PR TITLE
Introduce i64 Date annotation

### PR DESCRIPTION
**Pull request by kriskowal**
This PR adds a feature where an i64 can be annotated with js.type = 'Date'. The millisecond precision i64 timestamp will be read as a JavaScript Date with 53 bits precision. ISO timestamp strings, epoch milliseconds, and Date instances will all be coerced to the i64 milliseconds when writing.

With this change, TCurl will accept and display ISO timestamps anywhere an `i64 (js.type = 'Date')` is bought or sold.

r @malandrew @blampe @jcorbin

### CODE REVIEW

**Comment by Willyham** [comment]
:+1: docs?

**Comment by kriskowal** [comment]
@Willyham followed up with the docs update here.

r @abhinav

File: i64.js [hunk 1]
```diff
@@ -128,8 +127,32 @@ I64LongRW.prototype.readFrom = function readFrom(buffer, offset) {
...
+
+util.inherits(I64DateRW, I64RW);
+
+I64DateRW.prototype.readFrom = function readFrom(buffer, offset) {
+    var long = Long.fromBits(
+        buffer.readInt32BE(offset + 4, 4, true),
+        buffer.readInt32BE(offset + 0, 4, true)
+    );
+    var seconds = long.divide(1000).toNumber();
+    var value = new Date(seconds);
```
Comment (abhinav) [inline i64.js]: My understanding was `Date` takes milliseconds, not seconds?

Comment (abhinav) [inline i64.js]: Yeah. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date

File: i64.js [hunk 2]
```diff
@@ -128,8 +127,32 @@ I64LongRW.prototype.readFrom = function readFrom(buffer, offset) {
...
+    return new bufrw.ReadResult(null, offset + 8, value);
+};
+
+I64DateRW.prototype.writeInto = function writeInto(value, buffer, offset) {
+    var self = this;
+    if (typeof value === 'string') {
+        value = Date.parse(value);
```
Comment (abhinav) [inline i64.js]: Same. `.parse` returns the number of milliseconds since epoch.

Comment (kriskowal) [inline i64.js]: I will have a talk with my brain later about this. I have addressed the issue by getting rid of the scale machinations.

File: README.md [hunk 3]
```diff
@@ -365,20 +365,29 @@ typedef map<Struct, Struct> (js.type = 'entries') StructsByStructs
...
+- A `{hi, lo}` or `{high, low}` pair of 32 bit precision integers.
 - A 16 digit hexadecimal string, like `0102030405060708`.
 - An array of 8 byte values. Ranges are not checked, but coerced.
 - An 8 byte buffer.
 
-In a future version, we are likely to expose `js.type` annotations to read any
-of these forms off the wire.
-Currently, they can only be written.
+ThriftRW supports a type annotation for i64 that reifies i64 as an instance of
+[Long][], an object that models a 64 bit number as a `{high, low}` pair of 32
```
Comment (malandrew) [inline README.md]: is this supposed to be `[Long][]`? If so, what is meant by that annotation?

Comment (malandrew) [inline README.md]: nm. this looks like a markdown shorthand for a link.

Comment (kriskowal) [inline README.md]: Markdown autonomous link syntax.

**Comment by malandrew** [comment]
Other than @abhinav's parse comment, LGTM

**Pull request merged**